src/contract.rs

```diff
@@ -48,3 +57,125 @@ impl Contract {
         &mut self.state
     }
 }
+
+pub struct ContractInstrumenter<'a, S: Store> {
+    module: Module,
+    schedule: &'a crate::Schedule,
+    #[doc(hidden)]
+    _marker: PhantomData<S>,
+}
+
+impl<'a, S: Store> ContractInstrumenter<'a, S> {
+    /// Creates a new [`ContractInstrumenter`] instance from a mutable reference
+    /// to a [`Contract`] and a [`Schedule`] config.
+    pub(crate) fn instrument(
```

Comment:
Because I did not think at the moment that a trait would be a better solution and also on that way all of the instrumentalization code is self-contained in this struct impl block.